x/dex/keeper/tick_iterator.go

```diff
@@ -4,69 +4,61 @@ import (
 	"context"
 
 	"github.com/NicholasDotSol/duality/x/dex/types"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 type TickIterator struct {
-	start    int64
-	end      int64
-	pairId   *types.PairId
-	keeper   Keeper
-	ctx      sdk.Context
-	hasToken func(sdk.Context, *types.Tick) bool
-	nextTick func(tick int64) int64
-	stop     func(int64) bool
+	iter sdk.Iterator
+	cdc  codec.BinaryCodec
 }
 
-func (k Keeper) NewTickIterator(ctx context.Context,
+func (k Keeper) NewTickIterator(
+	// NOTE: both start and end are inclusive
+	ctx context.Context,
 	start int64,
 	end int64,
 	pairId *types.PairId,
-	findToken0 bool,
-	scanLeft bool) types.TickIteratorI {
-	var hasToken func(sdk.Context, *types.Tick) bool
-	var nextTick func(tick int64) int64
-	var stop func(int64) bool
-
-	if findToken0 {
-		hasToken = k.TickHasToken0
-	} else {
-		hasToken = k.TickHasToken1
-	}
+	scanLeft bool,
+	cdc codec.BinaryCodec,
+) types.TickIteratorI {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	prefixStore := prefix.NewStore(sdkCtx.KVStore(k.storeKey), types.TickPrefix(pairId))
 
 	if scanLeft {
-		nextTick = func(tick int64) int64 { tick--; return tick }
-		stop = func(idx int64) bool { return idx < end }
+		return TickIterator{
+			iter: prefixStore.ReverseIterator(
+				types.TickIndexToBytes(end),
+				types.TickIndexToBytes(start+1),
```

Comment:
This is a little unintuitive – start and end are both inclusive, but then we're rendering one exclusive on the basis of a 3rd parameter. Would seem better to have end always represent the "last element iterated" rather than the "right-most" element.

Reply:
Not following. By default The kvstore.Iterator has the end value as exclusive. This is less intuitive to me and doesn't line up with how we are interacting with the iterator. Ie. When we specific MaxTick we are using it as an inclusive boundary.